### PR TITLE
Summarise Logging

### PR DESCRIPTION
**Pull request by KoderKow**
In reference to issue #3 

Added summarize, summarize_at, summarize_all, summarize_each, summarize_if, and their summarise counterparts. Examples for summarize have been added to the README as well.

Let me know if the structure is good, I tried to keep it in line with your code.

Abstract Example:

``` r
mtcars %>%
    filter(mpg > 10) %>%
    group_by(cyl, vs, am) %>%
    summarise(total_weight = sum(wt)) %>%
    filter(total_weight > 3) %>%
    mutate(percent = total_weight/sum(total_weight)) %>%
    summarize(total_weight_sum = sum(total_weight)) %>%
    summarise_all(mean)
#> filter: no rows removed
#> group_by: 7 groups (cyl, vs, am)
#> summarise: 2 groups remaining (cyl, vs)
#> filter (grouped): removed 1 out of 7 rows (14%)
#> mutate (grouped): new variable 'percent' with 5 unique values and 0% NA
#> summarize: one group remaining (cyl)
#> summarise_all: 0 groups remaining
```

_Note:_ I had to use an if statement during the text output to avoid empty parentheses when there are no more groups remaining after a summarise. ie; "0 groups remaining ()". **format_list** does a great job of displaying the names when there are groups. Currently the ()'s are inside the glue statements. What do you think about possibly having the ()'s built into the **format_list** function to avoid this scenario?

### CODE REVIEW

File: R/summarize.R [hunk 1]
```diff
@@ -0,0 +1,92 @@
+#' Wrapper around dplyr::summarize and related functions
+#' that prints information about the operation
+#'
+#' @param .data a tbl; see \link[dplyr]{summarize}
+#' @param ... see \link[dplyr]{summarize}
+#' @return see \link[dplyr]{summarize}
+#' @import dplyr
+#' @export
+summarize <- function(.data, ...){
```
Comment (elbersb) [inline R/summarize.R]: can you add a space before the `{`? (here and in the other instances)

File: R/summarize.R [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+}
+
+log_summarize <- function(.data, fun, funname, ...){
+    newdata <- fun(.data, ...)
+
+    if (!"data.frame" %in% class(.data) | !should_display()) {
+        return(newdata)
+    }
+
+    group_names <- dplyr::groups(newdata)
```
Comment (elbersb) [inline R/summarize.R]: please use `get_groups` here (from tidylog.R) -- not sure if `dplyr::groups` works for all dplyr version

File: README.Rmd [hunk 3]
```diff
@@ -97,6 +97,22 @@ b <- full_join(band_members, band_instruments, by = "name")
 c <- anti_join(band_members, band_instruments, by = "name")
 ```
 
+### summarize
+
+```{r}
+a <- mtcars %>%
+  group_by(cyl, carb) %>%
```
Comment (elbersb) [inline README.Rmd]: please indent by 4 spaces -- as in the rest of the README

File: README.Rmd [hunk 4]
```diff
@@ -97,6 +97,22 @@ b <- full_join(band_members, band_instruments, by = "name")
...
+  summarize(total_weight = sum(wt))
+
+b <- mtcars %>%
+  group_by(cyl) %>%
+  summarise(total_weight = sum(wt))
```
Comment (elbersb) [inline README.Rmd]: I think we can drop this second example

**Comment by elbersb** [comment]
Good start! Looks like you're having fun ;) 
I put a few comments in the code. Your PR should also have new minimal tests for this functionality. Have a look in the tests folder to see examples.

How about we add some more information to summarize? For instance:
```r
#> summarize: now XX rows and XX columns, one group remaining (cyl)
```

and

```r
#> summarize: now XX rows and XX columns, 0 groups remaining
```

With format_list we should deal in a separate PR.

**Comment by KoderKow** [comment]
I addressed the reviews. Thank you for the patience and help, this is a lot of fun! I have learned a lot already.

Great idea on x rows and y columns, that has been added.

**Comment by elbersb** [comment]
Looks great. I think we should drop `summarize_each`/`summarise_each`, because they are deprecated (and we don't support `mutate_each` etc. either).

Are you going to add a test?

**Comment by KoderKow** [comment]
I have never done a test before, is below good? I tried looking at your tests as a reference.

``` r
library("dplyr")
library("tidylog")
context("test-summarize")

test_that("summarize", {
    expect_message({
        out <- summarize(mtcars, test = TRUE)
    })
    expect_equal(nrow(out), 1)
    
    expect_message({
        out <- mtcars %>%
            group_by(cyl, carb) %>%
            summarize(avg_mpg = mean(mpg))
    })
    expect_equal(nrow(out), 9)
    
    expect_message({
        out <- summarise_all(mtcars, max)
    })
    expect_equal(nrow(out), ncol(mtcars))
    
    expect_message({
        out <- summarise_at(mtcars, c("mpg", "hp"), mean)
    })
    
    expect_silent({
        out <- dplyr::summarise(mtcars)
    })
})

test_that("filter: scoped variants", {
    
})

test_that("filter: argument order", {
    expect_message({
        out <- summarize(mpg == 21, .data = mtcars)
    })
    expect_equal(nrow(out), 2)
})
```

File: R/summarize.R [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    log_summarize(.data, dplyr::summarize_at, "summarize_at", ...)
+}
+
+#' @rdname summarize
+#' @export
+summarize_if <- function(.data, ...) {
+    log_summarize(.data, dplyr::summarize_if, "summarize_if", ...)
+}
+
+#' Wrapper around dplyr::summarise and related functions
```
Comment (elbersb) [inline R/summarize.R]: I'd just let this point to `summarize` as well

File: tests/testthat/test-summarize.R [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+    expect_message({
+        out <- summarise_at(mtcars, c("mpg", "hp"), mean)
+    })
+
+    expect_silent({
+        out <- dplyr::summarise(mtcars)
+    })
+})
+
+test_that("filter: scoped variants", {
```
Comment (elbersb) [inline tests/testthat/test-summarize.R]: could you change this to `summarize` and move the tests for `summarise_all` and `summarise_at` here? (these are the scoped variants)

File: tests/testthat/test-summarize.R [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+})
+
+test_that("filter: argument order", {
```
Comment (elbersb) [inline tests/testthat/test-summarize.R]: filter -> summarize

**Comment by elbersb** [comment]
getting close! I added a few more comments.

Check this page to see which lines of your patch are covered by your tests: https://codecov.io/gh/elbersb/tidylog/compare/419fff677a54c90a25dae81702b68b6e57858bf7...f1ddbf33dd24446128ff70e904dc80186da02c2a/diff
Green lines are covered, red lines aren't. Not saying you have to add more tests (I'll do that soon for the whole project), but just letting you know that you can directly see which lines your tests cover.

File: tests/testthat/test-summarize.R (hunk 6 above)
Comment (KoderKow) [reply tests/testthat/test-summarize.R]: I moved them into the scope, thanks for being so helpful!

**Comment by KoderKow** [comment]
> getting close! I added a few more comments.
> 
> Check this page to see which lines of your patch are covered by your tests: https://codecov.io/gh/elbersb/tidylog/compare/419fff677a54c90a25dae81702b68b6e57858bf7...f1ddbf33dd24446128ff70e904dc80186da02c2a/diff
> Green lines are covered, red lines aren't. Not saying you have to add more tests (I'll do that soon for the whole project), but just letting you know that you can directly see which lines your tests cover.

Wow! That's so awesome! I still have a lot to learn about automated testing for packages! I believe I have added the fixes, let me know!

**Comment by elbersb** [comment]
great! just merged this.

**Pull request merged**